adaptdl/adaptdl/checkpoint.py

```diff
@@ -110,10 +110,17 @@ def save_state(state, sync=True):
         state.sync()
     if replica_rank() == 0:
         name = _STATES_TO_NAMES[state]
+        tmp_name = f"_{name}"
         if checkpoint_path() is not None:
-            with open(os.path.join(checkpoint_path(), name), "wb") as f:
+            ckpt_file = os.path.join(checkpoint_path(), name)
+            tmp_ckpt_file = os.path.join(checkpoint_path(), tmp_name)
+
+            with open(tmp_ckpt_file, "wb") as f:
                 state.save(f)
 
+            # Save checkpoint atomically to avoid file corruption
+            os.rename(tmp_ckpt_file, ckpt_file)
+
```

Comment:
I think this approach still has an issue when the process is killed in-between saving two state objects. In that case the overall checkpoint may consist of only partially-updated states. Better to create a sub-directory for each checkpoint I think, and delete the previous sub-directories.

Reply:
Makes sense. Checkpoints are saved by invoking `save_all_states()`, not individual `save()` function. I missed that. Thanks for pointing it out!

Reply:
Now the `save_state()` will write to a temp folder (`_checkpoint`). Once all states are saved, the temp folder will be renamed to `checkpoint`. Please review again.